### ISSUE TITLE
estimator.train() produces constant large loss (1e10) and NaNs on Xenium 5K data

### ISSUE THREAD

**Issue by IterateTillWin** [lugia-xiao/GITIII#3]
Hi,
thank you for releasing this great tool.

I am encountering an issue when running the model on my own Xenium 5K dataset. During the training step (estimator.train()), the loss is always extremely large (10000000000.0) and becomes NaN from the very beginning. After ~50 epochs, the training finishes but does not produce GRIT_best.pth, which is required for the next step estimator.calculate_influence_tensor().

Questions

**1. Is it normal for the loss to become NaN immediately when using custom Xenium 5K data?**
**2.  Are there any required preprocessing steps, normalization constraints, or value ranges that could cause this behavior if violated?**
Any guidance would be greatly appreciated.
Thank you very much for your help!

Below are the main observations and logs.

Start loading the dataset
Have samples: 1 ['test']
There are totally 8865 cells in this dataset
Finish loading the dataset
Train loader length: 28
Validation loader length: 7
Start training
Finish training, epoch: 0
Loss_all: nan ; Loss_not_interaction: nan median PCC all: tensor(nan) median PCC_not_interact: tensor(nan) max PCC all: tensor(nan) max PCC_not_interact: tensor(nan)
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
Finish validating, epoch: 0
Loss_all: nan ; Loss_not_interaction: nan median PCC all: tensor(nan) median PCC_not_interact: tensor(nan) max PCC all: tensor(nan) max PCC_not_interact: tensor(nan)
Best validation loss now: 10000000000.0
Finish training, epoch: 1
Loss_all: nan ; Loss_not_interaction: nan median PCC all: tensor(nan) median PCC_not_interact: tensor(nan) max PCC all: tensor(nan) max PCC_not_interact: tensor(nan)
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
torch.return_types.topk(
values=tensor([nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan, nan,
        nan, nan, nan, nan, nan, nan], device='cuda:0'),
indices=tensor([28, 29, 26, 27, 25, 24, 22, 23, 19, 18, 16, 17, 21, 20, 14, 15,  7,  6,
         4,  5,  1,  0,  2,  3, 11, 10,  8,  9, 13, 12], device='cuda:0'))
Finish validating, epoch: 1
Loss_all: nan ; Loss_not_interaction: nan median PCC all: tensor(nan) median PCC_not_interact: tensor(nan) max PCC all: tensor(nan) max PCC_not_interact: tensor(nan)
Best validation loss now: 10000000000.0
Finish training, epoch: 2

**Comment by Shipeng-Cheng** [comment]
I would like to add that the environment itself seems to be working correctly.
I was able to run the tutorial notebook tutorial_mouse_brain_MERFISH.nbconvert.ipynb with the provided example dataset without any issues, and the training completed normally with the expected output files generated.

Therefore, this issue seems to be specific to my custom Xenium 5K dataset, rather than an environment or installation problem.

**Comment by lugia-xiao** [comment]
Hi, thank you for your interest and for reporting this issue. You’re right—this is indeed a bug. While developing a new version of GITIII, I also noticed that the method can sometimes produce NaN values when running on datasets with a large number of genes, and I’ve fixed the problem.

I will update embedding.py accordingly. After reinstalling the package, the issue should be solved. Please let me know if you still encounter the problem.

**Comment by Shipeng-Cheng** [comment]
Thank you very much for the update and for fixing this issue so quickly. I have reinstalled the package with the updated embedding.py, and the method now works correctly without producing NaN values. The problem appears to be fully resolved on my datasets.
